fixed load/users getting distributed to missing worker

radhakrishnaakamat opened the pull request:
fixed load/users getting distributed to the missing workers. 
I am new to open source commit and python. 
Please let me know if anything needs to be changed.
fixes #2008 #2010

cyberw commented:
Can you add a unit test? You can override/shorten the timeout in the tests to reduce the time they take by using pytests mock feature (there should be some examples among the tests)

cyberw commented on locust/test/test_dispatch.py:
```diff
@@ -3273,6 +3273,29 @@ class User5(User):
+        users_dispatcher.new_dispatch(target_user_count=9, spawn_rate=3)
+        users_dispatcher._wait_between_dispatch = 0
+        users_dispatcher.prepare_rebalance()
+        list(users_dispatcher)
```
what does this line do? :)

radhakrishnaakamat replied:
re-organized the code. it validates the workers available for redistribution

cyberw commented on locust/test/test_dispatch.py:
```diff
@@ -3273,6 +3273,29 @@ class User5(User):
+        users_dispatcher._wait_between_dispatch = 0
+        users_dispatcher.prepare_rebalance()
+        list(users_dispatcher)
+        self.assertIs(len(users_dispatcher._worker_nodes), 2, "Expected worker 2 will be removed from worker nodes")
```
Can you also validate that the remaining workers get the right number of users?

radhakrishnaakamat replied:
added the test case

radhakrishnaakamat commented:
It asserts to Count of active workers in dispatch.. Here the missing worker
should be removed after heartdead interval.. Added comment to code and
added more sensible message in assert

On Sat, 12 Feb, 2022, 19:46 Lars Holmberg, ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In locust/test/test_dispatch.py
> <https://github.com/locustio/locust/pull/2010#discussion_r805165583>:
>
> > +            weight = 1
> +
> +        class User3(User):
> +            weight = 1
> +
> +        user_classes = [User1, User2, User3]
> +
> +        worker_nodes = [WorkerNode(str(i + 1)) for i in range(3)]
> +        worker_nodes[1].heartbeat = HEARTBEAT_DEAD - 1
> +
> +        users_dispatcher = UsersDispatcher(worker_nodes=worker_nodes, user_classes=user_classes)
> +
> +        users_dispatcher.new_dispatch(target_user_count=9, spawn_rate=3)
> +        users_dispatcher._wait_between_dispatch = 0
> +        users_dispatcher.prepare_rebalance()
> +        list(users_dispatcher)
>
> what does this line do? :)
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/locustio/locust/pull/2010#pullrequestreview-880859491>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AJXIJ6SAZXB7U6XXLQS2LHLU2ZTUVANCNFSM5OFSJBIA>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

cyberw commented on locust/dispatch.py:
```diff
@@ -16,6 +16,8 @@
 if TYPE_CHECKING:
     from locust.runners import WorkerNode
 
+HEARTBEAT_DEAD = -10
```
import from runners instead?

radhakrishnaakamat replied:
I tried that, but was getting error during initialization.
Will try the reverse tomorrow..

Could you help me with any resource on how to setup debugging in distributed mode.. That would be really helpful for me to contribute in a better way.. I am resorting to print statement for debugging(This is my first python project)

cyberw replied:
Sorry, I dont have much to contribute with debugging distributed. You know how to run an individual unit test, right?

radhakrishnaakamat replied:
Yes, initially used to run tox.. later found out the way to run individually.

cyberw commented on locust/runners.py:
```diff
@@ -59,6 +59,7 @@
 CPU_MONITOR_INTERVAL = 5.0
 HEARTBEAT_INTERVAL = 1
 HEARTBEAT_LIVENESS = 3
+HEARTBEAT_DEAD = -10
```
Feels weird that this is a negative number?

radhakrishnaakamat replied:
It symbolic to time with no response, as current heartbeat logic subtracts 1 for each missed heartbeat. Didn't want to introduce another variable to maintain skipped heart beat.

mboutet commented:
Hi @radhakrishnaakamat, can you link this PR to the corresponding issue #2008?

radhakrishnaakamat commented:
Fixes #<2010> #<2008>

radhakrishnaakamat commented:
@mboutet thanks for the pointers in #2008, I was able to fix it in the right places in a better way. 
@cyberw could you kindly review it once again, I have updated the tests and tox worked fine too. Thanks!!

cyberw commented:
👍👍
